site/content/en/guides/producer/functions/golang/_index.md

```diff
@@ -0,0 +1,314 @@
+)
+
+func main() {
+	type Data struct {
+		Value string `yaml:"value,omitempty"`
+	}
+	type Example struct {
+		Data Data `yaml:"data,omitempty"`
+	}
+	functionConfig := &Example{}
+
+	cmd := framework.Command(functionConfig, func(items []*yaml.RNode) ([]*yaml.RNode, error) {
+        // framework has parsed the input ResourceList.functionConfig into the functionConfig
+        // variable
+		for i := range items {
+			if err := items[i].PipeE(yaml.SetAnnotation("value", functionConfig.Data.Value)); err != nil {
+				return nil, err
+			}
+		}
+		return items, nil
+	})
+
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+}
+```
+
+Note: functionConfig need not read from the `data` field if it is not going to be run
+imperatively with `kpt fn run DIR/ --image gcr.io/some/image -- foo=bar` or 
+`kpt fn run DIR/ --exec-path /some/bin --enable-exec -- foo=bar`.  This is more appropriate
+for functions implementing abstractions (e.g. client-side CRD equivalents).
+
+```go
+...
+	type NestedValue struct {
+		Value string `yaml:"value,omitempty"`
+	}
+	type Spec struct {
+		NestedValue string `yaml:"nestedValue,omitempty"`
+        MapValues map[string]string  `yaml:"mapValues,omitempty"`
+        ListItems []string  `yaml:"listItems,omitempty"`
+	}
+	type Example struct {
+		Spec Spec `yaml:"spec,omitempty"`
+	}
+	functionConfig := &Example{}
+...
+```
+
+```yaml
+# PACKAGE_DIR/example.yaml
+apiVersion: example.com/v1alpha1
+kind: Example
+metadata:
+  name: foo
+  annotations:
+    config.kubernetes.io/function: |
+      exec:
+        path: /path/to/my-fn
+spec:
+  nestedValue:
+    value: something
+  mapValues:
+    key: value
+  listItems:
+  - a
+  - b
+```
+### kyaml
+
+Functions written in go should use the [sigs.k8s.io/kustomize/kyaml] libraries for modifying
+resource configuration.
+
+The [sigs.k8s.io/kustomize/kyaml/yaml] library offers utilities for reading and modifying
+yaml configuration, while retaining comments and structure.
+
+To use the kyaml/yaml library, become familiar with:
+ 
+- The `*yaml.RNode` type, which represents a configuration object or field
+  - [link](https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml/yaml?tab=doc#RNode)
+- The `Pipe` and `PipeE` functions, which apply a series of pipelined operations to the `*RNode`.
+  - [link](https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml/yaml?tab=doc#RNode.Pipe)
+
+
+#### Workflow
+
+To modify a *yaml.RNode call PipeE() on the *RNode, passing in the operations to be performed.
+
+```go
+// Set the spec.replicas field to 3 if it exists
+var node *yaml.RNode
+...
+err := node.PipeE(yaml.Lookup("spec", "replicas"), yaml.FieldSetter{StringValue: "3"})
+```
+
+```go
+// Set the spec.replicas field to 3, creating it if it doesn't exist
+var node *yaml.RNode
+...
+// pass in the type of the node to create if it doesn't exist (e.g. Sequence, Map, Scalar)
+err := node.PipeE(yaml.LookupCreate(yaml.ScalarNode, "spec", "replicas"), yaml.FieldSetter{StringValue: "3"})
```

Comment:
Related to our offline discussion. Reviewing the API here, I think it's still exposing low-level YAML AST constructs.

I'm not opposed to having this as a complementary API if people don't mind thinking in terms of YAML AST. But ideally the user should be thinking in terms of objects in the native language (Javascript objects or Go structs), and not in the terms of the low-level file format. I think we can get close to that reality by the orchestrator doing the heavy lifting in terms of preserving file format features like comments in YAML.

Reply:
Can you give an example of what this means?  `Unstructured` and `map[string]interface{}` do not seem like better alternatives to `yaml.Node`.  Providing both a high-level constructr as well as access to low-level constructs for break glass isn't a bad thing.

Reply:
Yea, unstructured in Go is also cumbersome to use. This is why we started with TS as the first SDK language.

Going back to the other comment, this is what I suggest: Present the different choice of languages and runtime (TS, Go/kyaml, star/python) with the same prominence (Same level in the hierarchy in hugo). They're at heterogeneous in terms of API and abstractions, but it's fine for now to get validation. This is in contrast to other tools (e.g. Pulumi and other SDKs) that attempt to provide the same abstractions in different languages.

Reply:
I am not sure i understand the issues you are concerned about.

With kyaml in go:

```go
v, err := value.Pipe(yaml.Lookup("path", "to", "value"))
```

With  lodash in js:

```js
v = _.get(value, "path", "to", "value")
```

Is the concern the package name: **yaml**.Lookup vs **resource**.Lookup?

Reply:
re: consistency across SDKs, I'd argue SDKs should stay as close to the IO spec as possible, with libraries offering value-adds on top.  e.g. using the nomenclature `functionConfig`, `ResourceList`, `results`, etc.  This will ensure consistency between the language SDKs.